Wangscape/image/TileRearranger.cpp

```diff
@@ -0,0 +1,263 @@
+#include "TileRearranger.h"
+#include <set>
+#include "logging/Logging.h"
+const static std::array<sf::Color, 6> default_region_colours{
+    sf::Color::Red,
+    sf::Color::Green,
+    sf::Color::Blue,
+    sf::Color::Yellow,
+    sf::Color::Cyan,
+    sf::Color::Magenta};
+
+TileRearranger::TileRearranger()
+{
+    for (int i = 0; i < Corners; i++)
+        regionColours[i] = default_region_colours[i];
+}
+
+const TileRearrangement TileRearranger::rearrangeTile(const sf::Image & base_tile,
+                                                      IVec offset_a,
+                                                      IVec offset_b)
+{
+    // Initialise vectors
+    mTileRearrangement.offsetA = offset_a;
+    mTileRearrangement.offsetB = offset_b;
+    mTileRearrangement.baseSize = makeUVec(base_tile.getSize());
+
+    // Check all region colours are unique
+    validateColours();
+    // Convert base tile image into stack of partition masks
+    decomposeBaseTile(base_tile);
+    // Check base tile regions are nonempty and nonoverlapping
+    validateBaseTile();
+
+    // Tessellate padded version of base tile in 3x3 grid 
+    const auto tessellation_offsets = makeTessellationOffsets();
+    const auto base_tessellation = tessellated(mTileRearrangement.base, tessellation_offsets);
+    // Check that tessellation has no holes or overlaps
+    validateTessellation(base_tessellation);
+
+    // Find bounding box of dual tile in base tile coordinates
+    calculateRearrangementParameters();
+    // Translate each region of the base tile partition into dual tile partition, and assemble them
+    makeDual();
+
+    // Tessellate padded version of dual tile in 3x3 grid
+    const auto dual_tessellation = tessellated(mTileRearrangement.dual, tessellation_offsets);
+    // Check that tessellation has no holes or overlaps
+    validateTessellation(dual_tessellation);
+
+    // Use adjacency between dual tile mask and offset versions to find edges and corners of dual tile mask
+    findDualBoundaries(dual_tessellation);
+
+    return mTileRearrangement;
+}
```

Comment:
You could achieve a lot with proper functions' naming instead of putting comments before each call. For example:
```cpp
validateColours(); // areUnique(regionColours);
decomposeBaseTile(base_tile); // convertToPartitionMasksStack(base_tile);
makeDual(); // translateToDualTileParition(baseParition); (and return dualPartition)
```